### DISCUSSION TITLE
Problems when use JSBSim in C++

### DISCUSSION THREAD

**Discussion by DreamerSurpass** [JSBSim-Team/jsbsim#887] [Q&A]
I want to write script in C++ to run JSBSim, just like Reference Manual.
![image](https://user-images.githubusercontent.com/93421761/231129626-6230a4da-dc33-452e-b2d1-e35858e09b3f.png)
But compiler cannot find FGFDMExec.h thought I have added the scr path.

I'm afraid I'm not familiar with C++. Any ideas? Thanks a lot.

**Comment by bcoconni** [comment]
I'd say your best course of action is to use the MSVC project files that are included with JSBSim. Open `JSBSim.sln` with Microsoft Visual Studio and compile the project. You should get an executable `JSBSim.exe`. From that point on, you could hack the file `src/JSBSim.cpp` to your liking and recompile the executable using MSVC.

**Reply by DreamerSurpass** [reply]
Thanks for your reply. I have used MSVC to build JSBSim, and I tested JSBSim.exe, it runs correctly. Now I want to build JSBSim.lib, and link it in my own code. I build a new Project and add src path and JSBSim.lib, then I run JSBSim_minimal.cpp, but when I instantiate FGFDMExec, there is a bug called "LNK2019 unsolved external symbol _declspec(dllimport)", how can I deal with this?

**Reply by DreamerSurpass** [reply]
I found if I annotating the code in JSBSim_API.h, and use JSBSim.lib in my own main.cpp, this error will not exist. 

Original: Add JSBSim.lib and run:
![image](https://user-images.githubusercontent.com/93421761/231401846-e86ff6f2-0c1f-4b69-a0b7-2e42f1b8dcb2.png)

After annotate the code in JSBSim_API.h, I can run my own project correctly:
![image](https://user-images.githubusercontent.com/93421761/231402093-f22cddae-6033-4fc2-9b20-922d07283f0e.png)

Any Ideas?

**Comment by seanmcleod** [comment]
The `JSBSIM_API` attributes are designed to be used when JSBSim is being built or used as a DLL. When the DLL is being built then `JSBSIM_EXPORT` will be defined which ensures that all the C++ classes are exported by the DLL. If a client app is using the JSBSim DLL then `JSBSIM_EXPORT` won't be defined which means `JSBSIM_API` will be expanded to `__declspec(dllimport)` basically telling the linker about the C++ classes in the DLL.

Now in your case if you've built JSBSim as a static lib and then you're trying to link to it as a static lib rather than use the DLL version then you must ensure that `JSBSIM_STATIC_LINK` is defined in your project, i.e. no need to go and comment things out in `JSBSim_API.h`, since in this case look at the `#else` section where `JSBSIM_API` is defined as nothing.

**Reply by DreamerSurpass** [reply]
Yes, I'm trying to link to it as a static lib rather than use DLL version,  but I don't quite understand how to define `JSBSIM_STATIC_LINK` in my project. Does this mean add Additional Dependencies in properties like this?
![image](https://user-images.githubusercontent.com/93421761/231418430-ee6ba735-93f0-4ff0-884a-7d63ab835cc9.png)

Or does this mean directly add define `JSBSIM_STATIC_LINK` in main.cpp code? Could you please give me a more specific explanation or a demo code demonstration? Thank you for your patient answer!

**Comment by seanmcleod** [comment]
No `Additional Dependencies` for the `Linker` are additional libraries, both regular static libs and import libs for DLLs that you want to link against.

The `#defines` in the code are processed by the pre-processor, before the compiler starts parsing the source code.

![image](https://user-images.githubusercontent.com/824856/231420836-cdc0000c-a4eb-487d-885b-59331fe70ffa.png)

> I'm afraid I'm not familiar with C++.

That's going to be tricky, since C++ isn't the easiest programming environment to get familiar with. You'll probably want to find some online tutorials/lessons that give you at least some of the basics.

**Reply by DreamerSurpass** [reply]
Thank you again for your patience. 
> That's going to be tricky, since C++ isn't the easiest programming environment to get familiar with. You'll probably want to find some online tutorials/lessons that give you at least some of the basics.

And thanks for your advice, I'll continue to learn.
